python/loyalty_workflow.py

```diff
@@ -0,0 +1,135 @@
+                send_email,
+                email_strings.EMAIL_WELCOME.format(self.customer.status.name),
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+
+        await workflow.wait_condition(
+            lambda: not self.customer.account_active or info.get_current_history_length() > EVENT_HISTORY_THRESHOLD
+        )
+
+        if self.customer.account_active:
+            logging.info(
+                "Account %s still active, but event history threshold reached; continuing-as-new." % self.customer.customerId)
+            workflow.continue_as_new(self.customer)
+
+        return "Loyalty workflow completed. Customer: %s" % self.customer.customerId
+
+    @workflow.signal(name=SIGNAL_CANCEL_ACCOUNT)
+    async def cancel_account(self) -> None:
+        self.customer.account_active = False
+        await workflow.execute_activity(
+                send_email,
+                email_strings.EMAIL_CANCEL_ACCOUNT,
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+
+    @workflow.signal(name=SIGNAL_ADD_POINTS)
+    async def add_points(self, points_to_add: int) -> None:
+        self.customer.points += points_to_add
+
+        status_change = self.customer.status.update(self.customer.points)
+        if status_change > 0:
+            await workflow.execute_activity(
+                send_email,
+                email_strings.EMAIL_PROMOTED.format(self.customer.status.name),
+                start_to_close_timeout=timedelta(seconds=5)
+            )
+        elif status_change < 0:
+            await workflow.execute_activity(
+                send_email,
+                email_strings.EMAIL_DEMOTED.format(self.customer.status.name),
+                start_to_close_timeout=timedelta(seconds=5)
+            )
+
+    @workflow.signal(name=SIGNAL_INVITE_GUEST)
+    async def invite_guest(self, guest_id: str) -> None:
+        if len(self.customer.guests) >= self.customer.status.guests_allowed:
+            await workflow.execute_activity(
+                send_email,
+                email_strings.EMAIL_INSUFFICIENT_POINTS,
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+            return
+
+        self.customer.guests.add(guest_id)
+
+        guest = Customer(customerId=guest_id)
+        child_workflow_id = CUSTOMER_WORKFLOW_ID_FORMAT.format(guest_id)
+        try:
+            child_handle = await workflow.start_child_workflow(
+                CustomerLoyaltyWorkflow.run,
+                guest,
+                id=child_workflow_id,
+                parent_close_policy=ParentClosePolicy.ABANDON,
+                id_reuse_policy=WorkflowIDReusePolicy.REJECT_DUPLICATE,
+            )
+        except WorkflowAlreadyStartedError:
+            logging.info("Child workflow already started")
+            child_handle: WorkflowHandle = workflow.get_external_workflow_handle(child_workflow_id)
+
+        child_info = await child_handle.describe()
```

Comment:
Do you have tests for this? You can't make a client call in a workflow. (Temporal has no facility for describing a workflow inside another workflow)

Reply:
I do now. This specific error made moot by moving this all to an activity.